main.go

```diff
@@ -326,10 +185,58 @@ func main() {
 	log.Printf("Took: " + time.Since(startTime).String())
 }
 
-func isSameStack(archiveStackID string, currentStackID string) bool {
+// Helpers
+
+// failf prints an error and terminates the step.
+func failf(format string, args ...interface{}) {
+	log.Errorf(format, args...)
+	os.Exit(1)
+}
+
+// parseArchiveInfo reads the stack id and architecture from the given json bytes.
+func parseArchiveInfo(b []byte) (info model.ArchiveInfo, err error) {
+	err = json.Unmarshal(b, &info)
+	return
+}
+
+func isSameStack(archiveStackInfo model.ArchiveInfo, currentStackInfo model.ArchiveInfo) bool {
 	// TODO This check is a temporary solution to support GEN2 VMs having different ids for same stack types
 	r := regexp.MustCompile("^(.+)-gen2.*$")
-	currentStackID = r.ReplaceAllString(currentStackID, "$1")
-	archiveStackID = r.ReplaceAllString(archiveStackID, "$1")
-	return archiveStackID == currentStackID
+	currentStackInfo.StackID = r.ReplaceAllString(currentStackInfo.StackID, "$1")
+	archiveStackInfo.StackID = r.ReplaceAllString(archiveStackInfo.StackID, "$1")
+
+	if archiveStackInfo.StackID != currentStackInfo.StackID {
+		return false
+	}
+
+	if isOldArchiveFormat(archiveStackInfo) {
+		return true
+	}
+
+	return archiveStackInfo.Architecture == currentStackInfo.Architecture
+}
+
+func isOldArchiveFormat(info model.ArchiveInfo) bool {
```

Comment:
I have trouble understanding what is considered "old" in this context and why we have a version number if we also check the architecture field.

Wouldn't it be enough to compare the version fields? My problem with this "old format" naming is that it makes sense at this point in time, sure, but what happens when we introduce another `ArchiveInfo` version in the future? What will "old" mean then, and how these fallback logic `if`s will look like in `isSameStack()`?

Reply:
@ofalvai I have included the architecture field just to be sure to check if it is empty but you are right that it is not needed. 

The definition of "old" will be always correct because it uses this:
```
func isOldArchiveFormat(info model.ArchiveInfo) bool {
	return info.Version < model.Version
}
```
so it compares the version number from the saved file with the exported constant version number coming from the package. If you update the format then you bump the version number there and this comparison will still be correct. 

After thinking about it what do you think about the following?

I will remove the `isOldArchiveFormat` function and modify the `main` to have this:
```
if archiveStackInfo.Version < model.Version {
    if archiveStackInfo.Architecture == "" {
            log.Warnf("Cache has missing architecture info so default (amd64) architecture is assumed")
    }

    log.Warnf("Please update your cache-push step to the latest version")
}
```

Reply:
This new code snippet looks a lot better! I don't want to waste time on hypothetical what-ifs either, I think the latest iteration is solid and we can always refactor it in the future if needed. Thanks for the clarification!